### ISSUE TITLE
FFTW windows build is not using SIMD

### ISSUE THREAD

**Issue by stevengj** [JuliaPackaging/Yggdrasil#534]
As discussed [on discourse](https://discourse.julialang.org/t/serious-slowdown-with-fftw-v1-1-0-and-v1-2-0-on-windows/34726), the Windows build is not using SIMD for some reason.

(My suspicion is that this is due to the build, not to anything in FFTW.jl.)

**Comment by giordano** [comment]
Inside the build environment for `x86_64-w64-mingw32`, before compiling I get:
```
sandbox:${WORKSPACE}/srcdir/fftw-3.3.9 # echo ${FLAGS[@]}
--prefix=/workspace/destdir --host=x86_64-w64-mingw32 --enable-shared --disable-static --disable-fortran --disable-mpi --disable-doc --enable-threads --with-combined-threads --enable-sse2 --enable-avx2 --with-our-malloc
```
`FLAGS` is then passed to `configure` in https://github.com/JuliaPackaging/Yggdrasil/blob/62d44097a26fe338763da8263b36ce6a63e7fa9c/F/FFTW/build_tarballs.jl#L59  This value should be correct, right?

**Comment by stevengj** [comment]
Yes, that looks correct.

**Comment by giordano** [comment]
I'm not at the computer now, but maybe you could check the log file in https://github.com/JuliaBinaryWrappers/FFTW_jll.jl/releases/download/FFTW-v3.3.9%2B3/FFTW.v3.3.9.x86_64-w64-mingw32.tar.gz to see if it has any hint about some features not being enabled?

**Comment by stevengj** [comment]
The logs seem to indicate that the SSE2 and AVX2 code is being compiled.  The only things I can think of are:

* FFTW checks at runtime whether the architecture supports SSE2 and AVX2.  Could those checks be failing for some reason, e.g. because binarybuilder messed with a CPUID instruction?  Hard to believe, since this doesn't happen on other platforms.  Also, FFTW [unconditionally enables SSE2 support on x86_64 systems](https://github.com/FFTW/fftw3/blob/c69d6c3c7512a814efe229d15e51dda0c1590ee6/simd-support/sse2.c#L32-L37), so at least that should be present.

* The arrays are not 16-byte aligned, so FFTW doesn't enable SIMD?  But that would cause earlier versions of FFTW.jl to enable SIMD as well, and reportedly that isn't happening.

@matteo-frigo, can you think of anything else that could be going wrong with our Windows cross-compile that could be disabling SIMD?  The build script is [here](https://github.com/JuliaPackaging/Yggdrasil/blob/db8d5675b3de01c8bcab9a7a224441e5f3d2ec07/F/FFTW/build_tarballs.jl#L14-L68).

**Comment by stevengj** [comment]
Another thing we are doing differently in the Windows build is [this line](https://github.com/JuliaPackaging/Yggdrasil/blob/db8d5675b3de01c8bcab9a7a224441e5f3d2ec07/F/FFTW/build_tarballs.jl#L48), which is apparently a workaround added by @staticfloat for [some problem](https://sourceforge.net/p/mingw-w64/mailman/message/36287627) in AVX2 support in MinGW gcc reported by @RoyiAvital, but it seems like that workaround succeeded?

**Comment by staticfloat** [comment]
Analyzing the instructions within the binary, it does indeed have both the CPUID instruction and AVX instructions within it.  I think someone with access to a windows machine and knowledge of FFTW internals is going to need to step through in a debugger to figure out what's going wrong here.

That workaround does seem very suspicious.  It's worthwhile seeing if we can build with a newer compiler and avoid using that workaround.

**Comment by matteo-frigo** [comment]
Try configuring fftw with --with-incoming-stack-boundary=2

At some point in the past, 32-bit gcc/mingw started assuming that the stack is 16-byte aligned, an assumption that was violated by various windows libraries.  The --with-incoming-stack-boundary=2 hack fixed the problem at least at that time.  I don't know what the current situation is, but it's worth a shot.

**Comment by stevengj** [comment]
We actually are using `--with-incoming-stack-boundary=2` on 32-bit Windows machines: https://github.com/JuliaPackaging/Yggdrasil/blob/db8d5675b3de01c8bcab9a7a224441e5f3d2ec07/F/FFTW/build_tarballs.jl#L44

Apparently, removing the old `-fno-asynchronous-unwind-tables` workaround fixed the problem (#536), though I have no idea why it would affect FFTW's ability to use SIMD instructions.

So it sounds like this issue can be closed, anyway.
